pkg/config/collector/collector.go

```diff
@@ -0,0 +1,372 @@
+	if err != nil {
+		return result, err
+	}
+
+	d, err := unstructured.ToYAML(stringToConfig(string(dat)))
+	if err != nil {
+		return result, err
+	}
+	err = yaml.Unmarshal(d, &result)
+
+	return result, err
+}
+
+func stringToConfig(s string) Config {
+	v := Config{}
+
+	splitted, _ := shlex.Split(s)
+	for _, item := range splitted {
+		parts := strings.SplitN(item, "=", 2)
+		value := "true"
+		if len(parts) > 1 {
+			value = strings.Trim(parts[1], `"`)
+		}
+		key := strings.Trim(parts[0], `"`)
+		v[key] = value
+	}
+
+	return v
+}
+
+// ConfigURL returns the value of config_url if set or empty string otherwise.
+func (c Config) ConfigURL() string {
+	if val, hasKey := c["config_url"]; hasKey {
+		if s, isString := val.(string); isString {
+			return s
+		}
+	}
+
+	return ""
+}
+
+func fetchRemoteConfig(url string) (*Config, error) {
+	var body []byte
+	result := &Config{}
+
+	err := retry.Do(
+		func() error {
+			resp, err := http.Get(url)
+			if err != nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusOK {
+				return fmt.Errorf("unexpected status: %d", resp.StatusCode)
+			}
+			defer resp.Body.Close()
+
+			body, err = io.ReadAll(resp.Body)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}, retry.Delay(time.Second), retry.Attempts(3),
+	)
+
+	if err != nil {
+		// TODO: improve logging
+		fmt.Println("could not fetch remote config: %w", err)
+		return result, nil
+	}
+
+	if !HasValidHeader(string(body)) {
+		// TODO: Print a warning when we implement proper logging
+		fmt.Println("No valid header in remote config: %w", err)
+		return result, nil
+	}
+
+	if err := yaml.Unmarshal(body, result); err != nil {
+		return result, fmt.Errorf("could not unmarshal remote config to an object: %w", err)
+	}
+
+	return result, nil
+}
+
+func HasValidHeader(data string) bool {
+	header := strings.SplitN(data, "\n", 2)[0]
+
+	// Trim trailing whitespaces
+	header = strings.TrimRightFunc(header, unicode.IsSpace)
+
+	// NOTE: we also allow "legacy" headers. Should only allow #cloud-config at
+	// some point.
+	return (header == DefaultHeader) || (header == "#kairos-config") || (header == "#node-config")
+}
+
+func (c Config) Query(s string) (res string, err error) {
+	s = fmt.Sprintf(".%s", s)
+
+	var dat map[string]interface{}
+
+	yamlStr, err := c.String()
+	if err != nil {
+		panic(err)
+	}
+	if err := yaml.Unmarshal([]byte(yamlStr), &dat); err != nil {
+		panic(err)
+	}
+
+	query, err := gojq.Parse(s)
+	if err != nil {
+		return res, err
+	}
+
+	iter := query.Run(dat) // or query.RunWithContext
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			break
+		}
+		if err, ok := v.(error); ok {
+			return res, fmt.Errorf("failed parsing, error: %w", err)
+		}
+
+		dat, err := yaml.Marshal(v)
+		if err != nil {
+			break
+		}
+		res += string(dat)
+	}
+	return
+}
+
+// TODO check if doing the right thing ... also checking remote files and cmdline.
+func FindYAMLWithKey(s string, opts ...Option) ([]string, error) {
```

Comment:
I can't find any references to this function. I think it was used before we were able to define the same block of configuration in 2 different files (deep merge). Back then, we just put the configuration in an existing file and this was a way to find which file defined the block. Now we simply create a new file (e.g. for the label mappings) so we don't need to lookup the other files.

It would still be useful if we wanted to make sure the same block is not defined in any other file (e.g. to make sure our file is the "effective" one) but afaik there is only one use-case (kcrypt) and that one always uses the same file.

Besides the above, it's also complex to return a sane return value when the block is defined in a remote `config_url`. What file do we return then? Even worse, when there is a chain of `config_urls` and the block is defined somewhere deeply in the chain.

I think we should remove this function. Please check that I didn't miss any references first.

Reply:
removed it